modules/standard/IO.chpl

```diff
@@ -3741,32 +3747,319 @@ proc channel.writeIt(const x) throws {
+    default this method will ignore leading whitespace when attempting to read
+    a literal.
+
+    :arg literal: the string to be matched.
+    :arg ignoreWhitespace: determines whether leading whitespace is ignored.
+
+    :throws BadFormatError: Thrown if literal could not be matched.
+    :throws EofError: Thrown if end of channel is encountered.
+
+  */
+  @unstable "channel.readLiteral is unstable and subject to change."
   inline
-  proc channel._writeLiteral(lit:string) throws {
-    var iolit = new ioLiteral(lit);
-    this.writeIt(iolit);
+  proc channel.readLiteral(literal:string,
+                           ignoreWhitespace=true) : void throws {
+    _readLiteralCommon(literal, ignoreWhitespace, isMatch=false);
+  }
+
+  /*
+    Advances the position of a channel by reading the exact bytes of the given
+    ``literal`` from the channel.
+
+    If the bytes are not matched exactly, then the channel's position is
+    unchanged. In such cases a :class:`OS.BadFormatError` will be thrown, unless the end of
+    the channel is encountered in which case an :class:`OS.EofError` will be thrown. By
+    default this method will ignore leading whitespace when attempting to read
+    a literal.
+
+    :arg literal: the bytes to be matched.
+    :arg ignoreWhitespace: determines whether leading whitespace is ignored.
+
+    :throws BadFormatError: Thrown if literal could not be matched.
+    :throws EofError: Thrown if end of channel is encountered.
+
+  */
+  @unstable "channel.readLiteral is unstable and subject to change."
+  inline
+  proc channel.readLiteral(literal:bytes,
+                           ignoreWhitespace=true) : void throws {
+    _readLiteralCommon(literal, ignoreWhitespace, isMatch=false);
   }
 
+  // TODO: Don't we need an option to ignore whitespace or not?
+  // Note: We don't want to allow skipping over non-whitespace.
+  //
+  // Note: We can add an 'ignoreWhitespace' optional argument that defaults
+  // to 'true' without changing behavior in existing programs.
   pragma "no doc"
   inline
-  proc channel._readNewline() throws {
+  proc channel._readNewlineCommon(param isMatch:bool) throws {
+    if writing {
+      param name = if isMatch then "matchNewline" else "readNewline";
+      compilerError(name + " on write-only channel", 2);
+    }
+
+    on this.home {
+      try! this.lock(); defer { this.unlock(); }
+      const err = qio_channel_skip_past_newline(false, _channel_internal,
+                                                /*skipWhitespaceOnly=*/ true);
+
+      const action = if isMatch then "matching" else "reading";
+      try _checkLiteralError("", err, action, isLiteral=false);
+    }
+  }
+
+  // non-unstable version we can use internally
+  pragma "no doc"
+  inline proc channel._readNewline() : void throws {
     var ionl = new ioNewline(true);
     this.readIt(ionl);
   }
 
+  // TODO: How does this differ from readln() ?
```

Comment:
readln will consume all of the line and throw it away. vs this will throw if the newline is not the next character.

Reply:
Gotcha.

And to clarify, I'm currently matching the behavior of ``ioNewline(true)``, which will skip leading whitespace. ``ioNewline(false)`` will skip all characters until a newline. So we don't have a way of throwing if the next character is not a newline at the moment.

Reply:
Right, I was talking about non-whitespace. The `readln()` would pass `abc\n` but AFAIK your new code will only pass `   \n` and the like. Agreed we will need some changes to handle `skipWhitespace=false` here.